Rules-Engine v2 - Targets Improvements

kennsippell opened the pull request:
# Description

This is the "Targets" work for the rules-engine v2. To be merged into the `tasks-v2` branch for AT.

- Stores target emissions in a _local pouchdb doc along with the contact state
- Aggregates the target emissions from the doc into targets
- Writes the aggregated targets into a doc which syncs to the server `target-2019-11-user_id`
- Disabled targets/tasks individually based on permissions
- Implements the isTimely window for tasks (60 days)
- On launch, starts a "freshness thread" which ensures all tasks and targets are up to date
- Changes the expected schema with partner code. Instead of a timestamp "startTime", we expect "readyStart" which is a number of days
- Some cleanup

#5657 
#5515

# Code review items

- Readable: Concise, well named, follows the [style guide](https://github.com/medic/medic-docs/blob/master/development/style-guide.md), documented if necessary.
- Documented: Configuration and user documentation on [medic-docs](https://github.com/medic/medic-docs/)
- Tested: Unit and/or e2e where appropriate
- Internationalised: All user facing text
- Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in Changes.md.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

garethbowen commented on shared-libs/rules-engine/src/refresh-rules-emissions.js:
```diff
@@ -11,39 +11,54 @@ const rulesEmitter = require('./rules-emitter');
-    const updatedTaskDocs = freshResults.filter(doc => doc.isUpdated).map(result => result.taskDoc);
-    return [...updatedTaskDocs, ...cancelledDocs];
-  });
+module.exports = (freshData = {}, calculationTimestamp = Date.now(), options = DEFAULT_OPTIONS) => {
```
Inline the default options and destructure properly, eg: `{ enableTasks=true, enableTargets=true }={}`.

garethbowen commented on shared-libs/rules-engine/src/rules-emitter.js:
```diff
@@ -67,7 +67,9 @@ module.exports = {
-    return ['startTime', 'endTime'].every(attr => Object.hasOwnProperty.call(Task.prototype, attr));
+    const Target = flow.getDefined('target');
+    const hasProperty = (obj, attr) => Object.hasOwnProperty.call(obj, attr);
+    return hasProperty(Task.prototype, 'readyStart') && hasProperty(Task.prototype, 'readyEnd') && hasProperty(Target.prototype, 'contact');
```
Too long - add line breaks after each `&&`.

garethbowen commented on shared-libs/rules-engine/src/rules-state-store.js:
```diff
@@ -177,12 +183,41 @@ const self = {
-    targets: settingsDoc && settingsDoc.targets,
+    targets: settingsDoc && settingsDoc.tasks && settingsDoc.tasks.targets,
+    can_view_tasks: settingsDoc && settingsDoc.permissions && settingsDoc.permissions.can_view_tasks,
+    can_view_analytics: settingsDoc && settingsDoc.permissions && settingsDoc.permissions.can_view_analytics,
```
There's a whole lot of repetition in here. What about...

```js
const settingsTasks = settingsDoc && settingsDoc.tasks || {};
const settingsPermissions = settingsDoc && settingsDoc.permissions || {};
const rulesConfig = {
  rules: settingsTasks.rules,
  targets: settingsTasks.targets,
  ...
  userDoc,
}
```

You could also do something clever with destructuring and Object.assign but I'll leave that as an exercise for the reader...

kennsippell replied:
No need for destructuring I don't think....

garethbowen commented on shared-libs/rules-engine/src/target-state.js:
```diff
@@ -0,0 +1,139 @@
+   *
+   * @param {Object} settingsDoc Settings document
+   */
+  empty: (settingsDoc) => {
```
I find the name of this function slightly misleading. What about `createEmptyState`?

kennsippell replied:
Agreed

garethbowen commented on shared-libs/rules-engine/src/task-states.js:
```diff
@@ -31,6 +45,25 @@ const States = {
+
+  const dueDate = moment(taskEmission.date);
+  return {
+    dueDate: dueDate.valueOf(),
```
Should this be `startOf('day')` too?

kennsippell replied:
i went for it

garethbowen commented on shared-libs/rules-engine/src/task-states.js:
```diff
@@ -31,6 +45,25 @@ const States = {
+  return {
+    dueDate: dueDate.valueOf(),
+    startTime: dueDate.clone().startOf('day').subtract(taskEmission.readyStart || 0, 'days').valueOf(),
+    endTime: dueDate.clone().startOf('day').add(taskEmission.readyEnd || 0, 'days').add(1, 'day').valueOf() - 1,
```
1. There's too much going on inline here and there's some duplication in date editing.
2. You're doing two additions to get the endTime which is unnecessary.
3. You don't need to clone the date as they are immutable - calling `add` returns a new moment object.
4. Should startTime and endTime be startDate and endDate respectively?
5. The `-1` at the end of the endTime calculation is probably unnecessary - that'll make it disappear 1ms earlier, right?

What about...

```js
const dueDate = moment(taskEmission.date).startOf('day');
const readyStart = taskEmission.readyStart || 0;
const readyEnd = (taskEmission.readyEnd || 0) + 1;
return {
  dueDate: dueDate.valueOf(),
  startDate: dueDate.subtract(readyStart, 'days').valueOf(),
  endDate: dueDate.add(readyEnd, 'days').valueOf(),
};
```

kennsippell replied:
Moments are not immutable and I believe the `clone` is necessary. https://momentjs.com/docs/#/manipulating/add/
> Mutates the original moment by adding time.

kennsippell replied:
> The -1 at the end of the endTime calculation is probably unnecessary - that'll make it disappear 1ms earlier, right?

If you take `moment().startOf('day').add(1, 'day')` you get a timestamp in tomorrow. 
If you subtract 1ms, you get a timestamp in today (11:59 pm).

kennsippell replied:
> Should startTime and endTime be startDate and endDate respectively?

Although this would be a decent description of the intent, I'd rather name the variable startTime since it is a timestamp.

kennsippell replied:
okay ya. i'll make that change

garethbowen replied:
> Moments are mutable

Wow, sorry, you're right, my memory failed me :(

garethbowen commented on webapp/src/js/controllers/analytics-targets.js:
```diff
@@ -12,19 +9,23 @@ angular.module('inboxControllers').controller('AnalyticsTargetsCtrl',
+    RulesEngine.isEnabled()
+      .then(isEnabled => {
+        ctrl.targetsDisabled = !isEnabled;
+        ctrl.loading = isEnabled;
```
This line isn't necessary - you set it to `false` in line 22.

garethbowen commented on webapp/src/js/controllers/analytics-targets.js:
```diff
@@ -12,19 +9,23 @@ angular.module('inboxControllers').controller('AnalyticsTargetsCtrl',
+      .catch(err => {
+        $log.error('Error getting targets', err);
         ctrl.loading = false;
+        ctrl.targets = [];
```
I'd move the `catch` up to line 21 and log the error and return `[]`. Then the `then` works like a `finally`. eg:

```js
  ...
  return isEnabled ? RulesEngine.fetchTargets() : [];
})
.catch(err => {
  $log.error('Error getting targets', err);
  return [];
})
.then(targets => {
  ctrl.targets = targets;
  ctrl.loading = false;
});
```

kennsippell replied:
:+1:

kennsippell commented on shared-libs/rules-engine/src/pouchdb-provider.js:
```diff
@@ -46,32 +43,41 @@ const medicPouchProvider = db => {
+      }
+
+      return commitTargetDocClosuers[_id](assign);
+    },
```
I saw conflict errors a few times during testing and decided to add this. 

There are several things which might result in either the `rulesStateStore` or this target doc being written twice quickly: users clicking quickly on different tabs, the "freshness" thread in the rules-engine service firing at the same time the user clicks the target tab, or the changes feed in the tasks controller firing at the same time the user clicks the tasks/targets tab. None of these scenarios are particularly likely. Avoiding conflicts in the state store is important, because if it gets out of sync then users see incorrect thigns. But conflicts in the target document isn't that important - just stale analytics data. So one could reasonable argue this is overkill and should just be removed.

kennsippell replied:
I think we should update the doc daily instead of on each view - which would also solve the conflicts for target docs.

garethbowen replied:
> I think we should update the doc daily instead of on each view - which would also solve the conflicts for target docs.

If I understand correctly this would mean if someone completed a task then the task completed count in the target wouldn't increment until midnight, right? We might be able to get away with this but it would need some consultation as it's a regression on the current responsive behaviour.

> There are several things which might result in either the rulesStateStore or this target doc being written twice quickly...

I think it would be cleaner if we had a debounce so if this was called twice the second call just returned the original promise rather than chaining up another one. I'm worried about it being slow so the user spamming the refresh button which just makes it slower...

kennsippell replied:
> If I understand correctly this would mean if someone completed a task then the task completed count in the target wouldn't increment until midnight, right?

We write target emissions into the `_local/rulesStateStore` document. We write these updates every time they happen. We avoid conflicts using existing code (`stateChangeCallback`). Not writing to this causes the symtomps you have described (target counts don't increment in UI).

This function is writing to the `target-2019-11` document which replicates for analytics purposes. It doesn't control any functionality of the webapp - it is an analytics-only document.

garethbowen replied:
Ahh, right. Yes that's fine. Would you use some kind of scheduler so it runs every day, or only update the doc if it's been more than a day since the last update, or...?

kennsippell replied:
Only update the doc if it has been more than a day since the last update + the existing [freshness thread](https://github.com/medic/cht-core/pull/6114/files#diff-8d5be28dba4aeadf015d6ae8168834d6R52) that runs once per launch. How does that sound?

garethbowen replied:
Sounds good. My only concern is that the doc could get a little out of date but the chance of someone running the app without restarts long enough is low enough that that's unlikely to happen.

kennsippell commented on shared-libs/rules-engine/src/provider-wireup.js:
```diff
@@ -0,0 +1,179 @@
+   * @param {Boolean=true} options.enableTasks Flag to enable tasks
+   * @param {Boolean=true} options.enableTargets Flag to enable targets
+   */
+  initialize: (provider, settingsDoc, userDoc, options) => {
```
Into it

kennsippell commented on shared-libs/rules-engine/src/provider-wireup.js:
```diff
@@ -0,0 +1,179 @@
+      .then(tasksToDisplay => {
+        const docsToCommit = updateTemporalStates(tasksToDisplay, calculationTimestamp);
+        provider.commitTaskDocs(docsToCommit);
+        return tasksToDisplay.filter(taskDoc => taskDoc.state === 'Ready');
```
Fixed. Scanned for other occurrences.

kennsippell commented:
@garethbowen Thanks for the great feedback. I've addressed your feedback + I've also added a mitigation users rewinding their clocks. It isn't perfect, but at least we don't create a new document each time the task is viewed -- which could cause a lot of problems for the user and the server.

garethbowen commented on shared-libs/rules-engine/src/pouchdb-provider.js:
```diff
@@ -46,32 +41,40 @@ const medicPouchProvider = db => {
+        user: userContactId,
+      });
+
+      const updateTimestamp = moment().startOf('day').valueOf();
```
I find this variable's name confusing - how about `const today`?

garethbowen commented on shared-libs/rules-engine/src/target-state.js:
```diff
@@ -0,0 +1,155 @@
+        return true;
+      }
+
+      const context = new Function('user', `return (${targetDefinition.context});`);
```
As discussed in Slack I think the old implementation using `$parse` is safer due to its more limited attack surface. I realise it makes the API more complicated because now you need to inject `$parse` somewhere, but I think it's worth it.

garethbowen commented on webapp/src/js/services/rules-engine.js:
```diff
@@ -42,14 +65,28 @@ angular.module('inboxServices').factory('RulesEngine', function(
+    items.forEach(item => {
+      item.isContextual = userDoc => item.context ? !!$parse(item.context)({ user: userDoc }) : true;
+    });
+    return settingsDoc;
```
I'm really concerned about modifying the extending Settings doc with functions like this. I get what you're trying to achieve but if I read this correctly it's modifying the cached settings object which is a bad idea. It would be a little better if you cloned the settings object but it's still mixing plain JSON with Functions in a dangerous way.

Can you not just inject the `$parse` service when you `RulesEngineCore.initialize`?

kennsippell replied:
I started with that and it just cluttered the interfaces quite a lot. I can revisit tomorrow.

kennsippell replied:
I've revisisted and `initialize()` now takes a structured settings object instead of the settingsDoc. This gives rules-engine angular service the ability to filter contextual targets. It simplifies the interface to rules-engine by combing "options" and "settings" into one structured parameter.
